### PR TITLE
[sil-devirtualizer] Use a more advanced algorithm to figure out the exact types of instances

### PR DESCRIPTION
**Pull request by swiftix**
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

- Look through BB arguments with multiple predecessors.
- Provide a new helper function to figure out the exact type of the underlying object. It will be used by subsequent commits to improve the escape analysis.

### CODE REVIEW

**Comment by swiftix** [comment]
@swift-ci Please smoke test

**Comment by swiftix** [comment]
@swift-ci Please smoke test Linux

File: lib/SILOptimizer/Utils/Devirtualize.cpp [hunk 1]
```diff
@@ -268,6 +268,143 @@ static SILValue getInstanceWithExactDynamicType(SILValue S, SILModule &M,
...
+/// type could not be determined.
+SILType swift::getExactDynamicType(SILValue S, SILModule &M,
+                                   ClassHierarchyAnalysis *CHA,
+                                   bool ForUnderlyingObject) {
+  // Set of values to be checked for their exact types.
+  SmallVector<SILValue, 8> WorkList;
+  // The detected type of the underlying object.
+  SILType ResultType;
+  // Set of processed values.
+  llvm::SetVector<SILValue> Processed;
```
Comment (eeckstein) [inline lib/SILOptimizer/Utils/Devirtualize.cpp]: No need to be a SetVector as you don't iterate over the elements

Comment (swiftix) [inline lib/SILOptimizer/Utils/Devirtualize.cpp]: Yep. I meant SmallSet, but typed SetVector ;-)

Comment (swiftix) [inline lib/SILOptimizer/Utils/Devirtualize.cpp]: @eeckstein I addressed it in https://github.com/apple/swift/pull/4761

**Pull request merged**